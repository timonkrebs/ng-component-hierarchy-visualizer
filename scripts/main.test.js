import { extractRoutesFromTS, flattenRoutes, resolveComponents, addDependencies, generateMermaid, main } from './helper'; // Adjust the path as needed

describe('extractRoutesFromTS', () => {
    it('should extract routes from TypeScript content', () => {
        const fileContent = `
            const routes: Routes = [
                { path: '', component: HomeComponent },
                { path: 'about', component: AboutComponent }
            ];
        `;
        const expectedRoutes = [
            { path: '', component: 'HomeComponent', parent: 'AppComponent' },
            { path: 'about', component: 'AboutComponent', parent: 'AppComponent' }
        ];
        const routes = extractRoutesFromTS(fileContent);
        expect(routes).toEqual(expectedRoutes);
    });
});

describe('flattenRoutes', () => {
    it('should flatten nested routes', () => {
        const nestedRoutes = [
            { path: '', component: 'HomeComponent', children: [
                { path: 'child', component: 'ChildComponent' }
            ]}
        ];
        const expectedRoutes = [
            { path: 'child', component: 'ChildComponent' },
            { path: '', component: 'HomeComponent', "children": null }
        ];
        const flatRoutes = flattenRoutes(nestedRoutes);
        expect(flatRoutes).toEqual(expectedRoutes);
    });
});

describe('resolveComponents', () => {
    it('should resolve components from routes', () => {
        process.env.INIT_CWD = "."
        const routes = [
            { path: '', component: 'HomeComponent' },
            { path: 'lazy', loadComponent: './lazy.component.ts' }
        ];
        const routesFileContent = `
            import { HomeComponent } from './home.component';
            import { LazyComponent } from './lazy.component';
        `;
        const expectedComponents = [
            { path: '', loadComponent: './home.component', componentType: 'HomeComponent', lazy: false, type: 'component' },
            { path: 'lazy', loadComponent: 'lazy.component.ts', lazy: true, type: 'component' }
        ];
        const components = resolveComponents(routes, routesFileContent);
        expect(components).toEqual(expectedComponents);
    });
});

describe('addDependencies', () => {
    it('should add dependencies to components', () => {
        const components = [
            { path: '', loadComponent: './home.component', componentType: 'HomeComponent', parent: 'AppComponent', lazy: false, type: 'component' }
        ];
        const expectedDependencies = [
            { path: '', loadComponent: './home.component', componentType: 'HomeComponent', parent: 'AppComponent', lazy: false, type: 'component' }
        ];
        const dependencies = addDependencies(components);
        expect(dependencies).toEqual(expectedDependencies);
    });
});

describe('generateMermaid', () => {
    it('should generate Mermaid diagram from routes', () => {
        const routes = [
            { path: '', componentType: 'HomeComponent', parent: 'AppComponent', type: 'component' }
        ];
        const expectedOutput = `
flowchart LR
AppComponent --> HomeComponent(HomeComponent)
        `.trim();
        expect(generateMermaid(routes)).toEqual(expectedOutput);
    });
});

describe('generateMain', () => {
    it('should resolve components from routes', () => {
        process.env.INIT_CWD = "./tests/route-definitions/ngx-admin";
        const routes = [
            { path: '', component: 'HomeComponent' },
            { path: 'lazy', loadComponent: './lazy.component.ts' }
        ];
        const routesFileContent = `
            import { HomeComponent } from './home.component';
            import { LazyComponent } from './lazy.component';
        `;
        const expectedComponents = [
            { path: '', loadComponent: './home.component', componentType: 'HomeComponent', lazy: false, type: 'component' },
            { path: 'lazy', loadComponent: 'lazy.component.ts', lazy: true, type: 'component' }
        ];
        const components = main('app-routing.module.ts');
        expect(components).toMatchSnapshot();
    });
});