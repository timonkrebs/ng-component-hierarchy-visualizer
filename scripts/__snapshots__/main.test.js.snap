// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDeepNestedRoutes should resolve components from routes 1`] = `
"flowchart LR
subgraph PagesModule
direction LR
pages --o ECommerceComponent(ECommerceComponent)
pages --o DashboardComponent(DashboardComponent)
subgraph LayoutModule
direction LR
layout --o StepperComponent(StepperComponent)
layout --o ListComponent(ListComponent)
layout --o InfiniteListComponent(InfiniteListComponent)
layout --o AccordionComponent(AccordionComponent)
layout --o Tab1Component(Tab1Component)
layout --o Tab2Component(Tab2Component)
layout --o TabsComponent(TabsComponent)
layout --o LayoutComponent(LayoutComponent)
end
pages -.-o LayoutModule(LayoutModule)
subgraph FormsModule
direction LR
forms --o FormInputsComponent(FormInputsComponent)
forms --o FormLayoutsComponent(FormLayoutsComponent)
forms --o ButtonsComponent(ButtonsComponent)
forms --o DatepickerComponent(DatepickerComponent)
forms --o FormsComponent(FormsComponent)
end
pages -.-o FormsModule(FormsModule)
subgraph UiFeaturesModule
direction LR
ui-features --o GridComponent(GridComponent)
ui-features --o IconsComponent(IconsComponent)
ui-features --o TypographyComponent(TypographyComponent)
ui-features --o SearchComponent(SearchComponent)
ui-features --o UiFeaturesComponent(UiFeaturesComponent)
end
pages -.-o UiFeaturesModule(UiFeaturesModule)
subgraph ModalOverlaysModule
direction LR
modal-overlays --o DialogComponent(DialogComponent)
modal-overlays --o WindowComponent(WindowComponent)
modal-overlays --o PopoversComponent(PopoversComponent)
modal-overlays --o TooltipComponent(TooltipComponent)
modal-overlays --o ToastrComponent(ToastrComponent)
modal-overlays --o ModalOverlaysComponent(ModalOverlaysComponent)
end
pages -.-o ModalOverlaysModule(ModalOverlaysModule)
subgraph ExtraComponentsModule
direction LR
extra-components --o CalendarComponent(CalendarComponent)
extra-components --o ProgressBarComponent(ProgressBarComponent)
extra-components --o SpinnerComponent(SpinnerComponent)
extra-components --o AlertComponent(AlertComponent)
extra-components --o CalendarKitFullCalendarShowcaseComponent(CalendarKitFullCalendarShowcaseComponent)
extra-components --o ChatComponent(ChatComponent)
extra-components --o ExtraComponentsComponent(ExtraComponentsComponent)
end
pages -.-o ExtraComponentsModule(ExtraComponentsModule)
subgraph MapsModule
direction LR
maps --o GmapsComponent(GmapsComponent)
maps --o LeafletComponent(LeafletComponent)
maps --o BubbleMapComponent(BubbleMapComponent)
maps --o SearchMapComponent(SearchMapComponent)
maps --o MapsComponent(MapsComponent)
end
pages -.-o MapsModule(MapsModule)
subgraph ChartsModule
direction LR
charts --o EchartsComponent(EchartsComponent)
charts --o D3Component(D3Component)
charts --o ChartjsComponent(ChartjsComponent)
charts --o ChartsComponent(ChartsComponent)
end
pages -.-o ChartsModule(ChartsModule)
subgraph EditorsModule
direction LR
editors --o TinyMCEComponent(TinyMCEComponent)
editors --o CKEditorComponent(CKEditorComponent)
editors --o EditorsComponent(EditorsComponent)
end
pages -.-o EditorsModule(EditorsModule)
subgraph TablesModule
direction LR
tables --o SmartTableComponent(SmartTableComponent)
tables --o TreeGridComponent(TreeGridComponent)
tables --o TablesComponent(TablesComponent)
end
pages -.-o TablesModule(TablesModule)
subgraph MiscellaneousModule
direction LR
miscellaneous --o NotFoundComponent(NotFoundComponent)
miscellaneous --o MiscellaneousComponent(MiscellaneousComponent)
end
pages -.-o MiscellaneousModule(MiscellaneousModule)
pages --o NotFoundComponent(NotFoundComponent)
pages --o PagesComponent(PagesComponent)
end
AppComponent -.-o PagesModule(PagesModule)
AppComponent --o NbLoginComponent(NbLoginComponent)
AppComponent --o NbRegisterComponent(NbRegisterComponent)
AppComponent --o NbLogoutComponent(NbLogoutComponent)
AppComponent --o NbRequestPasswordComponent(NbRequestPasswordComponent)
AppComponent --o NbResetPasswordComponent(NbResetPasswordComponent)
AppComponent --o NbAuthComponent(NbAuthComponent)
DashboardComponent --- NbThemeService{{NbThemeService}}
DashboardComponent --- SolarData{{SolarData}}
NotFoundComponent --- NbMenuService{{NbMenuService}}"
`;

exports[`generateDeepNestedRoutes should resolve template components from routes 1`] = `
"flowchart LR
subgraph PagesModule
direction LR
pages --o ECommerceComponent(ECommerceComponent)
pages --o DashboardComponent(DashboardComponent)
subgraph LayoutModule
direction LR
layout --o StepperComponent(StepperComponent)
layout --o ListComponent(ListComponent)
layout --o InfiniteListComponent(InfiniteListComponent)
layout --o AccordionComponent(AccordionComponent)
layout --o Tab1Component(Tab1Component)
layout --o Tab2Component(Tab2Component)
layout --o TabsComponent(TabsComponent)
layout --o LayoutComponent(LayoutComponent)
end
pages -.-o LayoutModule(LayoutModule)
subgraph FormsModule
direction LR
forms --o FormInputsComponent(FormInputsComponent)
forms --o FormLayoutsComponent(FormLayoutsComponent)
forms --o ButtonsComponent(ButtonsComponent)
forms --o DatepickerComponent(DatepickerComponent)
forms --o FormsComponent(FormsComponent)
end
pages -.-o FormsModule(FormsModule)
subgraph UiFeaturesModule
direction LR
ui-features --o GridComponent(GridComponent)
ui-features --o IconsComponent(IconsComponent)
ui-features --o TypographyComponent(TypographyComponent)
ui-features --o SearchComponent(SearchComponent)
ui-features --o UiFeaturesComponent(UiFeaturesComponent)
end
pages -.-o UiFeaturesModule(UiFeaturesModule)
subgraph ModalOverlaysModule
direction LR
modal-overlays --o DialogComponent(DialogComponent)
modal-overlays --o WindowComponent(WindowComponent)
modal-overlays --o PopoversComponent(PopoversComponent)
modal-overlays --o TooltipComponent(TooltipComponent)
modal-overlays --o ToastrComponent(ToastrComponent)
modal-overlays --o ModalOverlaysComponent(ModalOverlaysComponent)
end
pages -.-o ModalOverlaysModule(ModalOverlaysModule)
subgraph ExtraComponentsModule
direction LR
extra-components --o CalendarComponent(CalendarComponent)
extra-components --o ProgressBarComponent(ProgressBarComponent)
extra-components --o SpinnerComponent(SpinnerComponent)
extra-components --o AlertComponent(AlertComponent)
extra-components --o CalendarKitFullCalendarShowcaseComponent(CalendarKitFullCalendarShowcaseComponent)
extra-components --o ChatComponent(ChatComponent)
extra-components --o ExtraComponentsComponent(ExtraComponentsComponent)
end
pages -.-o ExtraComponentsModule(ExtraComponentsModule)
subgraph MapsModule
direction LR
maps --o GmapsComponent(GmapsComponent)
maps --o LeafletComponent(LeafletComponent)
maps --o BubbleMapComponent(BubbleMapComponent)
maps --o SearchMapComponent(SearchMapComponent)
maps --o MapsComponent(MapsComponent)
end
pages -.-o MapsModule(MapsModule)
subgraph ChartsModule
direction LR
charts --o EchartsComponent(EchartsComponent)
charts --o D3Component(D3Component)
charts --o ChartjsComponent(ChartjsComponent)
charts --o ChartsComponent(ChartsComponent)
end
pages -.-o ChartsModule(ChartsModule)
subgraph EditorsModule
direction LR
editors --o TinyMCEComponent(TinyMCEComponent)
editors --o CKEditorComponent(CKEditorComponent)
editors --o EditorsComponent(EditorsComponent)
end
pages -.-o EditorsModule(EditorsModule)
subgraph TablesModule
direction LR
tables --o SmartTableComponent(SmartTableComponent)
tables --o TreeGridComponent(TreeGridComponent)
tables --o TablesComponent(TablesComponent)
end
pages -.-o TablesModule(TablesModule)
subgraph MiscellaneousModule
direction LR
miscellaneous --o NotFoundComponent(NotFoundComponent)
miscellaneous --o MiscellaneousComponent(MiscellaneousComponent)
end
pages -.-o MiscellaneousModule(MiscellaneousModule)
pages --o NotFoundComponent(NotFoundComponent)
pages --o PagesComponent(PagesComponent)
end
AppComponent -.-o PagesModule(PagesModule)
AppComponent --o NbLoginComponent(NbLoginComponent)
AppComponent --o NbRegisterComponent(NbRegisterComponent)
AppComponent --o NbLogoutComponent(NbLogoutComponent)
AppComponent --o NbRequestPasswordComponent(NbRequestPasswordComponent)
AppComponent --o NbResetPasswordComponent(NbResetPasswordComponent)
AppComponent --o NbAuthComponent(NbAuthComponent)
DashboardComponent --- NbThemeService{{NbThemeService}}
DashboardComponent --- SolarData{{SolarData}}
NotFoundComponent --- NbMenuService{{NbMenuService}}"
`;

exports[`generateLazyComponents should resolve components from routes 1`] = `
"flowchart LR
AppComponent -.-o HomeComponent(HomeComponent)
AppComponent -.-o AuthComponent(AuthComponent)
AppComponent -.-o SettingsComponent(SettingsComponent)
subgraph ./features/profile/profile.routes
direction LR
profile -.-o ProfileArticlesComponent(ProfileArticlesComponent)
profile -.-o ProfileFavoritesComponent(ProfileFavoritesComponent)
profile --o ProfileComponent(ProfileComponent)
end
AppComponent -.-o ./features/profile/profile.routes(./features/profile/profile.routes)
AppComponent -.-o EditorComponent(EditorComponent)
AppComponent -.-o ArticleComponent(ArticleComponent)
HomeComponent --- TagsService{{TagsService}}
HomeComponent --- UserService{{UserService}}
UserService --- JwtService{{JwtService}}
AuthComponent --- UserService{{UserService}}
SettingsComponent --- UserService{{UserService}}
ProfileArticlesComponent --- ProfileService{{ProfileService}}
ProfileFavoritesComponent --- ProfileService{{ProfileService}}
ProfileComponent --- UserService{{UserService}}
ProfileComponent --- ProfileService{{ProfileService}}
EditorComponent --- ArticlesService{{ArticlesService}}
EditorComponent --- UserService{{UserService}}
ArticleComponent --- ArticlesService{{ArticlesService}}
ArticleComponent --- CommentsService{{CommentsService}}
ArticleComponent --- UserService{{UserService}}"
`;

exports[`generateLazyComponents should resolve template components from routes 1`] = `
"flowchart LR
AppComponent -.-o HomeComponent(HomeComponent)
HomeComponent ---ArticleListComponent([ArticleListComponent])
HomeComponent ---RxLet([RxLet])
HomeComponent ---IfAuthenticatedDirective([IfAuthenticatedDirective])
AppComponent -.-o AuthComponent(AuthComponent)
AuthComponent ---ListErrorsComponent([ListErrorsComponent])
AppComponent -.-o SettingsComponent(SettingsComponent)
SettingsComponent ---ListErrorsComponent([ListErrorsComponent])
subgraph ./features/profile/profile.routes
direction LR
profile -.-o ProfileArticlesComponent(ProfileArticlesComponent)
ProfileArticlesComponent ---ArticleListComponent([ArticleListComponent])
profile -.-o ProfileFavoritesComponent(ProfileFavoritesComponent)
ProfileFavoritesComponent ---ArticleListComponent([ArticleListComponent])
profile --o ProfileComponent(ProfileComponent)
ProfileComponent ---FollowButtonComponent([FollowButtonComponent])
end
AppComponent -.-o ./features/profile/profile.routes(./features/profile/profile.routes)
AppComponent -.-o EditorComponent(EditorComponent)
EditorComponent ---ListErrorsComponent([ListErrorsComponent])
AppComponent -.-o ArticleComponent(ArticleComponent)
ArticleComponent ---ArticleMetaComponent([ArticleMetaComponent])
ArticleComponent ---FollowButtonComponent([FollowButtonComponent])
ArticleComponent ---FavoriteButtonComponent([FavoriteButtonComponent])
ArticleComponent ---MarkdownPipe([MarkdownPipe])
ArticleComponent ---ListErrorsComponent([ListErrorsComponent])
ArticleComponent ---ArticleCommentComponent([ArticleCommentComponent])
ArticleComponent ---IfAuthenticatedDirective([IfAuthenticatedDirective])
HomeComponent --- TagsService{{TagsService}}
HomeComponent --- UserService{{UserService}}
UserService --- JwtService{{JwtService}}
AuthComponent --- UserService{{UserService}}
SettingsComponent --- UserService{{UserService}}
ProfileArticlesComponent --- ProfileService{{ProfileService}}
ProfileFavoritesComponent --- ProfileService{{ProfileService}}
ProfileComponent --- UserService{{UserService}}
ProfileComponent --- ProfileService{{ProfileService}}
EditorComponent --- ArticlesService{{ArticlesService}}
EditorComponent --- UserService{{UserService}}
ArticleComponent --- ArticlesService{{ArticlesService}}
ArticleComponent --- CommentsService{{CommentsService}}
ArticleComponent --- UserService{{UserService}}"
`;
